win dns records tests and Ansible.Basic use

Daniel-Sanchez-Fabregas opened the pull request:
I need this fixes to the tests in order to get them working.

Daniel-Sanchez-Fabregas committed:
win_dns_records tests: Ensure DNS services are installed

Daniel-Sanchez-Fabregas commented:
If you want to reproduce my integration test local environment setup for win_dns_record. This is the recipe:

~~~ bash
# Get Ansible repo
mkdir tmp
cd tmp
#git clone https://github.com/ansible/ansible
git clone https://github.com/johnboy2/ansible/pull/2
cd ansible/

# Get module to test
#git fetch origin pull/52317/head:pr_52317 # win_dns_record + patches
git fetch origin pull/2/head:pr_2
#git checkout pr_52317
git checkout pr_2

# Prepare environment to test
cp ./test/integration/inventory.winrm.template ./test/integration/inventory.winrm
vagrant init jborean93/WindowsServer2016
vagrant up

# Edit the inventory file
cat >./test/integration/inventory.winrm  <<\EOF
# This is the configuration template for ansible-test windows-integration tests.
# It can also be used with the legacy `make` based method of running tests.
#
# You do not need this template if you are:
#
# 1) Using the `--windows` option to provision temporary Windows instances on EC2.
#
# If you do not want to use the automatically provisioned temporary Windows instances,
# fill in the @VAR placeholders below and save this file without the .template extension.
#
# NOTE: Automatic provisioning of Windows instances on EC2 requires an ansible-core-ci API key.
#
# REMINDER: Standard ports for winrm are 5985 (HTTP) and 5986 (HTTPS).

# [windows]
# @NAME ansible_host=@HOST ansible_user=@USER ansible_password=@PASSWORD ansible_port=@PORT
#
# [windows:vars]
# ansible_connection=winrm
# ansible_winrm_server_cert_validation=ignore

[windows]
WindowsServer  ansible_host=127.0.0.1

[windows:vars]
ansible_user=vagrant
ansible_password=vagrant
ansible_port=55986
ansible_connection=winrm
ansible_winrm_transport=ntlm
ansible_winrm_server_cert_validation=ignore

# support winrm connection tests (temporary solution, does not support testing enable/disable of pipelining)
[winrm:children]
windows

# support tests that target testhost
[testhost:children]
windows
EOF

# Not in https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general_windows.html#developing-modules-general-windows page but in https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general.html#common-environment-setup
virtualenv venv
. venv/bin/activate
pip install -r requirements.txt
. hacking/env-setup

# Creating test
mkdir ./test/integration/targets/win_dns_record
# Creating test role
mkdir ./test/integration/targets/win_dns_record/tasks

vim ./test/integration/targets/win_dns_record/tasks/main.yml

pip install -r ./test/runner/requirements/windows-integration.txt
ansible-test windows-integration win_dns_record
~~~

johnboy2 commented on test/integration/targets/win_dns_record/tasks/main.yml:
```diff
@@ -4,6 +4,25 @@
     fail_on_missing: false
 
 - block:
+   - name: check if feature is availble
```
Can you explain your reasoning here? It looks like you want to guard against `win_feature` failing, and I'm not really sure why. To me, it looks like L16-24 alone would be fine.

Daniel-Sanchez-Fabregas replied:
You have a point. I just copied it from `win_rds/tasks/main.yml` test and seems not make sense here.

https://github.com/ansible/ansible/blob/devel/test/integration/targets/win_rds/tasks/main.yml#L4

Daniel-Sanchez-Fabregas committed:
win_dns_record partially converted to new Powershell module interface

win_dns_record converted to new Powershell module interface, except diff 
and required_if

Daniel-Sanchez-Fabregas committed:
win_dns_record: convert diff support

Daniel-Sanchez-Fabregas committed:
win_dns_record: convert diff support to after-before style

Daniel-Sanchez-Fabregas commented:
Conversion to new Powershell module interface solved.
Rebased to https://github.com/ansible/ansible/pull/51925

johnboy2 commented on lib/ansible/modules/windows/win_dns_record.ps1:
```diff
@@ -149,12 +141,13 @@ if ($values -ne $null -and $values.Count -gt 0)
+$records_end = Get-DnsServerResourceRecord -ZoneName $zone -Name $name -RRType $type -Node -ErrorAction:Ignore @extra_args | Sort-Object
 
-$module.Diff.prepared = ($changes -Join '')
+$module.Diff.before = @($records | ForEach-Object { "[$zone] $($_.HostName) $($_.TimeToLive.TotalSeconds) $type $($_.RecordData.$record_argument_name.ToString())`n" }) -join ''
```
These changes still don't work right in check mode -- because it should show what _would_ have been changed [(reference)](https://docs.ansible.com/ansible/2.7/user_guide/playbooks_checkmode.html#showing-differences-with-diff).

This diffing method may be elegant, but it's also demonstrably wrong.

Daniel-Sanchez-Fabregas replied:
Ouch! 
I wanted to get rid of `prepared` (not appearing in [2.8 docs](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_general_windows.html#windows-new-module-development)) and didn't realize.

johnboy2 replied:
Okay, that makes sense -- the lack of `prepared` seems a likely victim of the move away from the various legacy methods. That said, we'll still need to figure a way for diff-mode to be accurate when in check-mode. I'll take another stab at it today or tomorrow.

Daniel-Sanchez-Fabregas committed:
Don't test for Add-WindowsFeature

Daniel-Sanchez-Fabregas committed:
win_dns_record: Fix diff

When check mode is used diff changes must be simulated.

Daniel-Sanchez-Fabregas commented:
Diff should be OK now @johnboy2
